src/Data/Array/Polarized/Push.hs

```diff
@@ -11,55 +10,113 @@
+-------------------------------------------------------------------------------
 
--- XXX: the use of Vector in the type of alloc is temporary (see also
--- "Data.Array.Destination")
 -- | Convert a push array into a vector by allocating. This would be a common
 -- end to a computation using push and pull arrays.
 alloc :: Array a %1-> Vector a
-alloc (Array k n) = DArray.alloc n (k id)
+alloc (Array k) = allocArrayWriter $ k singletonWriter where
+  singletonWriter :: a -> ArrayWriter a
+  singletonWriter a = ArrayWriter (DArray.fill a) 1
+
+  allocArrayWriter :: ArrayWriter a %1-> Vector a
+  allocArrayWriter (ArrayWriter writer len) = DArray.alloc len writer
 
 -- | @`make` x n@ creates a constant push array of length @n@ in which every
 -- element is @x@.
-make :: a -> Int -> Array a
-make x n = Array (\k -> DArray.replicate (k x)) n
+make :: HasCallStack => a -> Int -> Array a
+make x n
+  | n < 0 = error "Making negative length push array"
+  | otherwise = Array (\makeA -> mconcat $ Prelude.replicate n (makeA x))
+
+
+-- # Instances
+-------------------------------------------------------------------------------
+
+instance Data.Functor Array where
+  fmap f (Array k) = Array (\g -> k (\x -> (g (f x))))
+
+instance Prelude.Semigroup (Array a) where
+  (<>) x y = append x y
+
+instance Semigroup (Array a) where
+  (<>) = append
+
+instance Prelude.Monoid (Array a) where
+  mempty = empty
+
+instance Monoid (Array a) where
+  mempty = empty
+
+empty :: Array a
+empty = Array (\_ -> mempty)
 
--- | Concatenate two push arrays.
 append :: Array a %1-> Array a %1-> Array a
-append (Array kl nl) (Array kr nr) =
-    Array
-      (\f dest -> parallelApply f kl kr (DArray.split nl dest))
-      (nl+nr)
-  where
-    parallelApply :: (a %1-> b) -> ((a %1-> b) -> DArray b %1-> ()) %1-> ((a %1-> b) -> DArray b %1-> ()) %1-> (DArray b, DArray b) %1-> ()
-    parallelApply f' kl' kr' (dl, dr) = kl' f' dl <> kr' f' dr
+append (Array k1) (Array k2) = Array (\writeA -> k1 writeA <> k2 writeA)
+
+instance Prelude.Semigroup (ArrayWriter a) where
+  (<>) x y = addWriters x y
+
+instance Prelude.Monoid (ArrayWriter a) where
+  mempty = emptyWriter
+
+instance Semigroup (ArrayWriter a) where
+  (<>) = addWriters
+
+instance Monoid (ArrayWriter a) where
+  mempty = emptyWriter
+
+addWriters :: ArrayWriter a %1-> ArrayWriter a %1-> ArrayWriter a
+addWriters (ArrayWriter k1 l1) (ArrayWriter k2 l2) =
+  ArrayWriter
+    (\darr ->
+      (DArray.split l1 darr) & \(darr1,darr2) -> consume (k1 darr1, k2 darr2))
+    (l1+l2)
+-- Remark. In order for the function above to work, consume must forcibly
+-- evaluate both tuples. If it was lazy, then we might not actually perform
+-- @k1@ or @k2@ and the unsafe IO won't get done. In general, this makes me
+-- think we haven't spelled out the careful rules of what consuming functions
+-- must follow so that we can release a safe API.
+
+emptyWriter :: ArrayWriter a
+emptyWriter = ArrayWriter (Unsafe.toLinear (const ())) 0
```

Comment:
Unsafe call here assumes that we can just forget about the `DArray`. I think this is the case; but it might worth a comment here. (I think every unsafe call deserves a comment unless incredibly obvious.)

Reply:
There shouldn't be an `Unsafe.toLinear` in this module. If we are missing a way to consume an empty destination array, then let's add it to the destination array module.

Reply:
I'll make an unsafe empty `DArray` consumer.